### PR TITLE
Cores per process for hipercow_parallel

### PR DESCRIPTION
**Pull request by weshinsley**

### CODE REVIEW

File: R/parallel.R [hunk 1]
```diff
@@ -110,49 +150,60 @@ hipercow_parallel_set_cores <- function(cores, envir = NULL) {
...
-hipercow_parallel_setup <- function(method) {
-  if (is.null(method)) {
+hipercow_parallel_setup <- function(parallel) {
+  if (is.null(parallel$method)) {
     return()
   }
 
-  cores <- hipercow_parallel_get_cores()
-  if (is.na(cores)) {
+  processes <- hipercow_parallel_get_cores() %/% parallel$cores_per_process
```
Comment (richfitz) [inline R/parallel.R:158]: Should we be printing something here to the logs if this is wasteful?

File: R/parallel.R [hunk 2]
```diff
@@ -3,47 +3,87 @@
...
 ##' @export
-hipercow_parallel <- function(method = NULL) {
+hipercow_parallel <- function(method = NULL, cores_per_process = 1L) {
+  assert_scalar_integer(cores_per_process)
+
   if (!is.null(method) &&
       (!method %in% c("future", "parallel"))) {
     cli::cli_abort(c(
-      "Parallel method {method} unknown.",
+      'Parallel method "{method}" unknown.',
```
Comment (richfitz) [inline R/parallel.R:82]:
```suggestion
      "Parallel method '{method}' unknown",
```

Normally we quote this way around, and no full stops

File: tests/testthat/test-parallel.R [hunk 3]
```diff
@@ -1,6 +1,6 @@
 test_that("Validate parallel args", {
   expect_error(hipercow_parallel("potato"),
-               "Parallel method potato unknown")
+               'Parallel method "potato" unknown')
```
Comment (richfitz) [inline tests/testthat/test-parallel.R:3]: This changes slightly with the above suggestion

File: R/parallel.R [hunk 4]
```diff
@@ -3,47 +3,87 @@
...
+##' process when launching a local cluster using one of the parallel
+##' methods. By default, this will be `1`. See details.
+##'
 ##'
 ##' @return A list containing your parallel configuration.
 ##'
 ##' @export
-hipercow_parallel <- function(method = NULL) {
+hipercow_parallel <- function(method = NULL, cores_per_process = 1L) {
+  assert_scalar_integer(cores_per_process)
```
Comment (richfitz) [inline R/parallel.R:77]: positive integer?

File: R/parallel.R [hunk 5]
```diff
@@ -110,49 +150,60 @@ hipercow_parallel_set_cores <- function(cores, envir = NULL) {
...
+           "each with {cores_per_process} core{?s}"))
 
   # rscript_libs is already set by default to .libPaths() in future::plan
   # but we also want to call set cores...
 
-  future::plan(future::multisession, workers = cores,
-               rscript_startup = "hipercow::hipercow_parallel_set_cores(1)")
+  future::plan(future::multisession, workers = processes,
+               rscript_startup = sprintf(
+                 "hipercow::hipercow_parallel_set_cores(%s)",
```
Comment (richfitz) [inline R/parallel.R:186]:
```suggestion
                 "hipercow::hipercow_parallel_set_cores(%d)",
```

**Review by richfitz — CHANGES_REQUESTED** [review]

File: R/parallel.R (hunk 1 above)
Comment (weshinsley) [reply R/parallel.R]: Added a cli_alert_info

File: R/parallel.R (hunk 2 above)
Comment (weshinsley) [reply R/parallel.R]: Done

File: R/parallel.R (hunk 4 above)
Comment (weshinsley) [reply R/parallel.R]: Done - I'll leave large numbers for the next PR on validation

File: R/parallel.R (hunk 5 above)
Comment (weshinsley) [reply R/parallel.R]: Done

**Review by richfitz — APPROVED** [review]

**Pull request merged**